### PR TITLE
Fixed crash on frequent video item change

### PR DESCRIPTION
**Pull request by nasser-munshi**
1. Fixed crash on frequent video item change for KVO add/remove mismatch (used Notification instead)
2. Fixed scrubber playing problem on video switch

### CODE REVIEW

**Comment by Krisiacik** [comment]
Thanks for the efforts @nasser-munshi ! 👍

@zfoltin pls let me check this before merge in the next couple of days. I want to make sure it fixes the issue but more importantly it doesn't inadvertently change desired UX.

File: ImageViewer/Source/GalleryViewController.swift [hunk 1]
```diff
@@ -589,17 +589,13 @@ open class GalleryViewController: UIPageViewController, ItemControllerDelegate {
 
     public func itemControllerWillAppear(_ controller: ItemController) {
 
-        if let videoController = controller as? VideoViewController {
-
-            scrubber.player = videoController.player
-        }
     }
```
Comment (RuiAAPeres) [inline ImageViewer/Source/GalleryViewController.swift]: Does this method need to be here? Empty implementation

File: ImageViewer/Source/MediaPlayer.swift [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    public struct Notification {
+        public static var status: NSNotification.Name {
+            return NSNotification.Name(rawValue: "media-player-status-notif")
+        }
+        public static var rate: NSNotification.Name {
+            return NSNotification.Name(rawValue: "media-player-rate-notif")
+        }
+    }
+    
+    public let avPlayer: AVPlayer
```
Comment (RuiAAPeres) [inline ImageViewer/Source/MediaPlayer.swift]: It's a property, so please move it to the top

File: ImageViewer/Source/MediaPlayer.swift [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    
+    public init(avPlayer: AVPlayer) {
+        self.avPlayer = avPlayer
+        super.init()
+        
+        //observer to avplayer properties
+        self.avPlayer.addObserver(self, forKeyPath: "status", options: NSKeyValueObservingOptions.new, context: nil)
+        self.avPlayer.addObserver(self, forKeyPath: "rate", options: NSKeyValueObservingOptions.new, context: nil)
+
```
Comment (RuiAAPeres) [inline ImageViewer/Source/MediaPlayer.swift]: Extra line

File: ImageViewer/Source/MediaPlayer.swift [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    }
+    
+    deinit {
+        //Remove observers
+        self.avPlayer.removeObserver(self, forKeyPath: "status")
+        self.avPlayer.removeObserver(self, forKeyPath: "rate")
+
```
Comment (RuiAAPeres) [inline ImageViewer/Source/MediaPlayer.swift]: Extra line

File: ImageViewer/Source/MediaPlayer.swift [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        NotificationCenter.default.post(name: Notification.status, object: avPlayer)
+    }
+    
+    func changePlayerRate() {
+        NotificationCenter.default.post(name: Notification.rate, object: avPlayer)
+    }
+    
+    //-------------------------------
+    //MARK: - KVO
+    //-------------------------------
```
Comment (RuiAAPeres) [inline ImageViewer/Source/MediaPlayer.swift]: Remove this comment.

File: ImageViewer/Source/MediaPlayer.swift [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    
+    override public func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
+        
+        if keyPath == "rate" {
+            changePlayerRate()
+        } else if keyPath == "status" {
+            changePlayerStatus()
+        }
+    }
+
```
Comment (RuiAAPeres) [inline ImageViewer/Source/MediaPlayer.swift]: Extra line

File: ImageViewer/Source/VideoScrubber.swift [hunk 7]
```diff
@@ -21,44 +22,46 @@ open class VideoScrubber: UIControl {
...
+//            if newValue == nil {
 
-                if let player = player {
+                if let player = mediaPlayer {
 
                     ///KVO
-                    player.removeObserver(self, forKeyPath: "status")
-                    player.removeObserver(self, forKeyPath: "rate")
+//                    player.removeObserver(self, forKeyPath: "status")
+//                    player.removeObserver(self, forKeyPath: "rate")
```
Comment (RuiAAPeres) [inline ImageViewer/Source/VideoScrubber.swift]: Remove commented code.

File: ImageViewer/Source/VideoScrubber.swift [hunk 8]
```diff
@@ -21,44 +22,46 @@ open class VideoScrubber: UIControl {
...
 
                     ///TIMER
                     if let periodicObserver = self.periodicObserver {
 
-                        player.removeTimeObserver(periodicObserver)
+                        player.avPlayer.removeTimeObserver(periodicObserver)
                         self.periodicObserver = nil
                     }
+                    
+
```
Comment (RuiAAPeres) [inline ImageViewer/Source/VideoScrubber.swift]: Extra lines

File: ImageViewer/Source/VideoScrubber.swift [hunk 9]
```diff
@@ -21,44 +22,46 @@ open class VideoScrubber: UIControl {
...
+                
                 ///NC
                 NotificationCenter.default.addObserver(self, selector: #selector(didEndPlaying), name: NSNotification.Name.AVPlayerItemDidPlayToEndTime, object: nil)
-
+                
+                NotificationCenter.default.addObserver(self, selector: #selector(playerChanged), name: MediaPlayer.Notification.rate, object: player.avPlayer)
+                NotificationCenter.default.addObserver(self, selector: #selector(playerChanged), name: MediaPlayer.Notification.status, object: player.avPlayer)
+                
+                
                 ///TIMER
```
Comment (RuiAAPeres) [inline ImageViewer/Source/VideoScrubber.swift]: Extra lines

File: ImageViewer/Source/VideoView.swift [hunk 10]
```diff
@@ -61,21 +60,18 @@ class VideoView: UIView {
...
             if status == .readyToPlay && rate != 0 {
-
+                
                 UIView.animate(withDuration: 0.3, animations: { [weak self] in
-
+                    
                     if let strongSelf = self {
-
+                        
                         strongSelf.previewImageView.alpha = 0
```
Comment (RuiAAPeres) [inline ImageViewer/Source/VideoView.swift]: These changes in the indentation should be removed.

File: ImageViewer/Source/VideoViewController.swift [hunk 11]
```diff
@@ -8,199 +8,198 @@
...
     init(index: Int, itemCount: Int, fetchImageBlock: @escaping FetchImageBlock, videoURL: URL, scrubber: VideoScrubber, configuration: GalleryConfiguration, isInitialController: Bool = false) {
-
+        
         self.videoURL = videoURL
         self.scrubber = scrubber
-        self.player = AVPlayer(url: self.videoURL)
-
+        self.mediaPlayer = MediaPlayer(avPlayer: AVPlayer(url: self.videoURL))
+        
         super.init(index: index, itemCount: itemCount, fetchImageBlock: fetchImageBlock, configuration: configuration, isInitialController: isInitialController)
```
Comment (RuiAAPeres) [inline ImageViewer/Source/VideoViewController.swift]: Remove all these changes in the indentation.

**Pull request closed without merging**